File: sdks/android/daten/src/main/kotlin/com/daten/sdk/DatenClient.kt
/**
 * DatenClient.kt
 * Daten Android SDK
 *
 * Native Android SDK for Document Management System
 * Version: 3.3.0
 */

package com.daten.sdk

import android.content.Context
import android.content.SharedPreferences
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import java.util.*
import java.util.concurrent.TimeUnit

// MARK: - Configuration

data class DatenConfiguration(
    val baseUrl: String,
    val apiKey: String? = null,
    val enableOfflineMode: Boolean = true,
    val enableBiometrics: Boolean = true,
    val syncIntervalSeconds: Long = 300
)

// MARK: - Models

data class Document(
    val id: String = UUID.randomUUID().toString(),
    var title: String,
    var content: String,
    val createdAt: Date = Date(),
    var updatedAt: Date = Date(),
    val userId: String,
    var version: Int = 1
)

data class AuthCredentials(
    val username: String,
    val password: String
)

data class AuthToken(
    val accessToken: String,
    val refreshToken: String,
    val expiresIn: Int,
    val tokenType: String
)

// MARK: - Sync Status

sealed class SyncStatus {
    object Idle : SyncStatus()
    object Syncing : SyncStatus()
    object Completed : SyncStatus()
    data class Failed(val error: Throwable) : SyncStatus()
}

// MARK: - Errors

sealed class DatenException(message: String) : Exception(message) {
    object NotAuthenticated : DatenException("User is not authenticated")
    data class NetworkError(val cause: Throwable) : DatenException("Network error: ${cause.message}")
    object InvalidResponse : DatenException("Invalid server response")
    object BiometricAuthFailed : DatenException("Biometric authentication failed")
    object OfflineModeDisabled : DatenException("Offline mode is disabled")
    object SyncConflict : DatenException("Sync conflict detected")
}

// MARK: - Main SDK Client

class DatenClient(
    private val context: Context,
    private val configuration: DatenConfiguration
) {

    // MARK: - Public State Flows

    private val _isAuthenticated = MutableStateFlow(false)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated.asStateFlow()

    private val _currentUser = MutableStateFlow<String?>(null)
    val currentUser: StateFlow<String?> = _currentUser.asStateFlow()

    private val _syncStatus = MutableStateFlow<SyncStatus>(SyncStatus.Idle)
    val syncStatus: StateFlow<SyncStatus> = _syncStatus.asStateFlow()

    private val _documents = MutableStateFlow<List<Document>>(emptyList())
    val documents: StateFlow<List<Document>> = _documents.asStateFlow()

    // MARK: - Private Properties

    private var authToken: AuthToken? = null
    private val httpClient: OkHttpClient
    private val prefs: SharedPreferences
    private var syncJob: Job? = null
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    // MARK: - Services

    private val authService: AuthenticationService by lazy { AuthenticationService(this) }
    private val documentService: DocumentService by lazy { DocumentService(this) }
    private val syncManager: SyncManager by lazy { SyncManager(this) }
    private val biometricService: BiometricAuthService by lazy { BiometricAuthService(context) }

    init {
        httpClient = OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .build()

        prefs = context.getSharedPreferences("daten_sdk", Context.MODE_PRIVATE)

        loadAuthToken()

        if (configuration.enableOfflineMode) {
            setupAutoSync()
        }
    }

    // MARK: - Authentication

    suspend fun signIn(credentials: AuthCredentials): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            val token = authService.signIn(credentials)
            authToken = token
            _isAuthenticated.value = true
            _currentUser.value = credentials.username
            saveAuthToken(token)

            if (configuration.enableOfflineMode) {
                syncManager.performSync()
            }

            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    fun signOut() {
        authToken = null
        _isAuthenticated.value = false
        _currentUser.value = null
        _documents.value = emptyList()
        clearAuthToken()
        stopAutoSync()
    }

    suspend fun refreshToken(): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            val currentToken = authToken ?: throw DatenException.NotAuthenticated

            val newToken = authService.refreshToken(currentToken.refreshToken)
            authToken = newToken
            saveAuthToken(newToken)

            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // MARK: - Biometric Authentication

    suspend fun authenticateWithBiometrics(
        activity: FragmentActivity,
        title: String = "Authenticate",
        subtitle: String = "Unlock Daten"
    ): Result<Boolean> {
        if (!configuration.enableBiometrics) {
            return Result.success(false)
        }

        return biometricService.authenticate(activity, title, subtitle)
    }

    // MARK: - Document Operations

    suspend fun fetchDocuments(): Result<List<Document>> = withContext(Dispatchers.IO) {
        try {
            if (!_isAuthenticated.value) {
                throw DatenException.NotAuthenticated
            }

            val docs = documentService.fetchDocuments()
            _documents.value = docs
            Result.success(docs)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun createDocument(document: Document): Result<Document> = withContext(Dispatchers.IO) {
        try {
            if (!_isAuthenticated.value) {
                throw DatenException.NotAuthenticated
            }

            val created = documentService.createDocument(document)

            if (configuration.enableOfflineMode) {
                syncManager.queueOperation(SyncOperation.Create(created))
            }

            _documents.value = _documents.value + created
            Result.success(created)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun updateDocument(document: Document): Result<Document> = withContext(Dispatchers.IO) {
        try {
            if (!_isAuthenticated.value) {
                throw DatenException.NotAuthenticated
            }

            val updated = documentService.updateDocument(document)

            if (configuration.enableOfflineMode) {
                syncManager.queueOperation(SyncOperation.Update(updated))
            }

            _documents.value = _documents.value.map {
                if (it.id == document.id) updated else it
            }

            Result.success(updated)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun deleteDocument(id: String): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            if (!_isAuthenticated.value) {
                throw DatenException.NotAuthenticated
            }

            documentService.deleteDocument(id)

            if (configuration.enableOfflineMode) {
                syncManager.queueOperation(SyncOperation.Delete(id))
            }

            _documents.value = _documents.value.filter { it.id != id }
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // MARK: - Sync Operations

    suspend fun syncNow(): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            if (!configuration.enableOfflineMode) {
                throw DatenException.OfflineModeDisabled
            }

            _syncStatus.value = SyncStatus.Syncing

            syncManager.performSync()

            _syncStatus.value = SyncStatus.Completed
            Result.success(Unit)
        } catch (e: Exception) {
            _syncStatus.value = SyncStatus.Failed(e)
            Result.failure(e)
        }
    }

    // MARK: - Private Helpers

    private fun setupAutoSync() {
        syncJob = scope.launch {
            while (isActive) {
                delay(configuration.syncIntervalSeconds * 1000)
                if (_isAuthenticated.value) {
                    try {
                        syncNow()
                    } catch (e: Exception) {
                        // Log error but continue syncing
                    }
                }
            }
        }
    }

    private fun stopAutoSync() {
        syncJob?.cancel()
        syncJob = null
    }

    private fun saveAuthToken(token: AuthToken) {
        prefs.edit()
            .putString("access_token", token.accessToken)
            .putString("refresh_token", token.refreshToken)
            .putInt("expires_in", token.expiresIn)
            .putString("token_type", token.tokenType)
            .apply()
    }

    private fun loadAuthToken() {
        val accessToken = prefs.getString("access_token", null)
        val refreshToken = prefs.getString("refresh_token", null)
        val expiresIn = prefs.getInt("expires_in", 0)
        val tokenType = prefs.getString("token_type", "Bearer")

        if (accessToken != null && refreshToken != null) {
            authToken = AuthToken(accessToken, refreshToken, expiresIn, tokenType!!)
            _isAuthenticated.value = true
        }
    }

    private fun clearAuthToken() {
        prefs.edit().clear().apply()
    }

    // MARK: - Internal API

    internal suspend fun <T> makeRequest(
        endpoint: String,
        method: String = "GET",
        body: String? = null,
        parser: (String) -> T
    ): T = withContext(Dispatchers.IO) {
        val url = "${configuration.baseUrl}$endpoint"

        val requestBuilder = Request.Builder().url(url)

        when (method) {
            "GET" -> requestBuilder.get()
            "POST" -> requestBuilder.post(
                body?.toRequestBody("application/json".toMediaType())
                    ?: "".toRequestBody()
            )
            "PUT" -> requestBuilder.put(
                body?.toRequestBody("application/json".toMediaType())
                    ?: "".toRequestBody()
            )
            "DELETE" -> requestBuilder.delete()
        }

        requestBuilder.addHeader("Content-Type", "application/json")

        authToken?.let {
            requestBuilder.addHeader("Authorization", "Bearer ${it.accessToken}")
        } ?: configuration.apiKey?.let {
            requestBuilder.addHeader("X-API-Key", it)
        }

        val request = requestBuilder.build()

        val response = httpClient.newCall(request).execute()

        if (!response.isSuccessful) {
            throw DatenException.NetworkError(IOException("HTTP ${response.code}"))
        }

        val responseBody = response.body?.string() ?: throw DatenException.InvalidResponse

        parser(responseBody)
    }

    fun close() {
        stopAutoSync()
        scope.cancel()
        httpClient.dispatcher.executorService.shutdown()
        httpClient.connectionPool.evictAll()
    }
}

// MARK: - Authentication Service

internal class AuthenticationService(private val client: DatenClient) {

    suspend fun signIn(credentials: AuthCredentials): AuthToken {
        val json = JSONObject().apply {
            put("username", credentials.username)
            put("password", credentials.password)
        }

        return client.makeRequest("/auth/login", "POST", json.toString()) { response ->
            val jsonResponse = JSONObject(response)
            AuthToken(
                accessToken = jsonResponse.getString("access_token"),
                refreshToken = jsonResponse.getString("refresh_token"),
                expiresIn = jsonResponse.getInt("expires_in"),
                tokenType = jsonResponse.getString("token_type")
            )
        }
    }

    suspend fun refreshToken(refreshToken: String): AuthToken {
        val json = JSONObject().apply {
            put("refresh_token", refreshToken)
        }

        return client.makeRequest("/auth/refresh", "POST", json.toString()) { response ->
            val jsonResponse = JSONObject(response)
            AuthToken(
                accessToken = jsonResponse.getString("access_token"),
                refreshToken = jsonResponse.getString("refresh_token"),
                expiresIn = jsonResponse.getInt("expires_in"),
                tokenType = jsonResponse.getString("token_type")
            )
        }
    }
}

// MARK: - Document Service

internal class DocumentService(private val client: DatenClient) {

    suspend fun fetchDocuments(): List<Document> {
        return client.makeRequest("/documents") { response ->
            val jsonArray = org.json.JSONArray(response)
            (0 until jsonArray.length()).map { i ->
                val obj = jsonArray.getJSONObject(i)
                Document(
                    id = obj.getString("id"),
                    title = obj.getString("title"),
                    content = obj.getString("content"),
                    createdAt = Date(obj.getLong("created_at")),
                    updatedAt = Date(obj.getLong("updated_at")),
                    userId = obj.getString("user_id"),
                    version = obj.getInt("version")
                )
            }
        }
    }

    suspend fun createDocument(document: Document): Document {
        val json = documentToJson(document)

        return client.makeRequest("/documents", "POST", json.toString()) { response ->
            jsonToDocument(JSONObject(response))
        }
    }

    suspend fun updateDocument(document: Document): Document {
        val json = documentToJson(document)

        return client.makeRequest("/documents/${document.id}", "PUT", json.toString()) { response ->
            jsonToDocument(JSONObject(response))
        }
    }

    suspend fun deleteDocument(id: String) {
        client.makeRequest<Unit>("/documents/$id", "DELETE") { }
    }

    private fun documentToJson(document: Document): JSONObject {
        return JSONObject().apply {
            put("id", document.id)
            put("title", document.title)
            put("content", document.content)
            put("created_at", document.createdAt.time)
            put("updated_at", document.updatedAt.time)
            put("user_id", document.userId)
            put("version", document.version)
        }
    }

    private fun jsonToDocument(json: JSONObject): Document {
        return Document(
            id = json.getString("id"),
            title = json.getString("title"),
            content = json.getString("content"),
            createdAt = Date(json.getLong("created_at")),
            updatedAt = Date(json.getLong("updated_at")),
            userId = json.getString("user_id"),
            version = json.getInt("version")
        )
    }
}

// MARK: - Sync Manager

internal sealed class SyncOperation {
    data class Create(val document: Document) : SyncOperation()
    data class Update(val document: Document) : SyncOperation()
    data class Delete(val id: String) : SyncOperation()
}

internal class SyncManager(private val client: DatenClient) {
    private val pendingOperations = mutableListOf<SyncOperation>()

    fun queueOperation(operation: SyncOperation) {
        synchronized(pendingOperations) {
            pendingOperations.add(operation)
        }
    }

    suspend fun performSync() {
        val operations = synchronized(pendingOperations) {
            pendingOperations.toList()
        }

        for (operation in operations) {
            when (operation) {
                is SyncOperation.Create -> client.documentService.createDocument(operation.document)
                is SyncOperation.Update -> client.documentService.updateDocument(operation.document)
                is SyncOperation.Delete -> client.documentService.deleteDocument(operation.id)
            }
        }

        synchronized(pendingOperations) {
            pendingOperations.clear()
        }
    }
}

// MARK: - Biometric Service

internal class BiometricAuthService(private val context: Context) {

    fun canAuthenticateWithBiometrics(): Boolean {
        val biometricManager = androidx.biometric.BiometricManager.from(context)
        return biometricManager.canAuthenticate(androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_STRONG) ==
                androidx.biometric.BiometricManager.BIOMETRIC_SUCCESS
    }

    suspend fun authenticate(
        activity: FragmentActivity,
        title: String,
        subtitle: String
    ): Result<Boolean> = suspendCancellableCoroutine { continuation ->

        if (!canAuthenticateWithBiometrics()) {
            continuation.resume(Result.failure(DatenException.BiometricAuthFailed)) {}
            return@suspendCancellableCoroutine
        }

        val executor = ContextCompat.getMainExecutor(context)

        val biometricPrompt = BiometricPrompt(
            activity,
            executor,
            object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                    continuation.resume(Result.success(true)) {}
                }

                override fun onAuthenticationFailed() {
                    continuation.resume(Result.failure(DatenException.BiometricAuthFailed)) {}
                }

                override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                    continuation.resume(Result.failure(DatenException.BiometricAuthFailed)) {}
                }
            }
        )

        val promptInfo = BiometricPrompt.PromptInfo.Builder()
            .setTitle(title)
            .setSubtitle(subtitle)
            .setNegativeButtonText("Cancel")
            .build()

        biometricPrompt.authenticate(promptInfo)
    }
}

File: sdks/ios/Daten/Sources/DatenClient.swift
///
/// DatenClient.swift
/// Daten iOS SDK
///
/// Native iOS SDK for Document Management System
/// Version: 3.3.0
///

import Foundation
import Combine
import LocalAuthentication

// MARK: - SDK Configuration

public struct DatenConfiguration {
    public let baseURL: URL
    public let apiKey: String?
    public let enableOfflineMode: Bool
    public let enableBiometrics: Bool
    public let syncInterval: TimeInterval

    public init(
        baseURL: URL,
        apiKey: String? = nil,
        enableOfflineMode: Bool = true,
        enableBiometrics: Bool = true,
        syncInterval: TimeInterval = 300
    ) {
        self.baseURL = baseURL
        self.apiKey = apiKey
        self.enableOfflineMode = enableOfflineMode
        self.enableBiometrics = enableBiometrics
        self.syncInterval = syncInterval
    }
}

// MARK: - Document Model

public struct Document: Codable, Identifiable {
    public let id: String
    public var title: String
    public var content: String
    public var createdAt: Date
    public var updatedAt: Date
    public var userId: String
    public var version: Int

    public init(
        id: String = UUID().uuidString,
        title: String,
        content: String,
        createdAt: Date = Date(),
        updatedAt: Date = Date(),
        userId: String,
        version: Int = 1
    ) {
        self.id = id
        self.title = title
        self.content = content
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.userId = userId
        self.version = version
    }
}

// MARK: - Authentication

public struct AuthCredentials {
    public let username: String
    public let password: String

    public init(username: String, password: String) {
        self.username = username
        self.password = password
    }
}

public struct AuthToken: Codable {
    public let accessToken: String
    public let refreshToken: String
    public let expiresIn: Int
    public let tokenType: String
}

// MARK: - Sync Status

public enum SyncStatus {
    case idle
    case syncing
    case completed
    case failed(Error)
}

// MARK: - SDK Errors

public enum DatenError: LocalizedError {
    case notAuthenticated
    case networkError(Error)
    case invalidResponse
    case biometricAuthFailed
    case offlineModeDisabled
    case syncConflict

    public var errorDescription: String? {
        switch self {
        case .notAuthenticated:
            return "User is not authenticated"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        case .invalidResponse:
            return "Invalid server response"
        case .biometricAuthFailed:
            return "Biometric authentication failed"
        case .offlineModeDisabled:
            return "Offline mode is disabled"
        case .syncConflict:
            return "Sync conflict detected"
        }
    }
}

// MARK: - Main SDK Client

@MainActor
public class DatenClient: ObservableObject {

    // MARK: - Published Properties

    @Published public private(set) var isAuthenticated = false
    @Published public private(set) var currentUser: String?
    @Published public private(set) var syncStatus: SyncStatus = .idle
    @Published public private(set) var documents: [Document] = []

    // MARK: - Private Properties

    private let configuration: DatenConfiguration
    private var authToken: AuthToken?
    private var cancellables = Set<AnyCancellable>()
    private let urlSession: URLSession
    private var syncTimer: Timer?

    // MARK: - Services

    private lazy var authService = AuthenticationService(client: self)
    private lazy var documentService = DocumentService(client: self)
    private lazy var syncManager = SyncManager(client: self)
    private lazy var biometricService = BiometricAuthService()

    // MARK: - Initialization

    public init(configuration: DatenConfiguration) {
        self.configuration = configuration

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 300
        self.urlSession = URLSession(configuration: config)

        loadAuthToken()

        if configuration.enableOfflineMode {
            setupAutoSync()
        }
    }

    // MARK: - Authentication

    public func signIn(credentials: AuthCredentials) async throws {
        let token = try await authService.signIn(credentials: credentials)
        self.authToken = token
        self.isAuthenticated = true
        self.currentUser = credentials.username
        saveAuthToken(token)

        // Start syncing
        if configuration.enableOfflineMode {
            try await syncManager.performSync()
        }
    }

    public func signOut() {
        authToken = nil
        isAuthenticated = false
        currentUser = nil
        documents = []
        clearAuthToken()
        stopAutoSync()
    }

    public func refreshToken() async throws {
        guard let currentToken = authToken else {
            throw DatenError.notAuthenticated
        }

        let newToken = try await authService.refreshToken(refreshToken: currentToken.refreshToken)
        self.authToken = newToken
        saveAuthToken(newToken)
    }

    // MARK: - Biometric Authentication

    public func authenticateWithBiometrics() async throws -> Bool {
        guard configuration.enableBiometrics else {
            return false
        }

        return try await biometricService.authenticate(reason: "Unlock Daten")
    }

    // MARK: - Document Operations

    public func fetchDocuments() async throws -> [Document] {
        guard isAuthenticated else {
            throw DatenError.notAuthenticated
        }

        let docs = try await documentService.fetchDocuments()
        self.documents = docs
        return docs
    }

    public func createDocument(_ document: Document) async throws -> Document {
        guard isAuthenticated else {
            throw DatenError.notAuthenticated
        }

        let created = try await documentService.createDocument(document)

        if configuration.enableOfflineMode {
            await syncManager.queueOperation(.create(created))
        }

        self.documents.append(created)
        return created
    }

    public func updateDocument(_ document: Document) async throws -> Document {
        guard isAuthenticated else {
            throw DatenError.notAuthenticated
        }

        let updated = try await documentService.updateDocument(document)

        if configuration.enableOfflineMode {
            await syncManager.queueOperation(.update(updated))
        }

        if let index = documents.firstIndex(where: { $0.id == document.id }) {
            documents[index] = updated
        }

        return updated
    }

    public func deleteDocument(id: String) async throws {
        guard isAuthenticated else {
            throw DatenError.notAuthenticated
        }

        try await documentService.deleteDocument(id: id)

        if configuration.enableOfflineMode {
            await syncManager.queueOperation(.delete(id))
        }

        documents.removeAll(where: { $0.id == id })
    }

    // MARK: - Sync Operations

    public func syncNow() async throws {
        guard configuration.enableOfflineMode else {
            throw DatenError.offlineModeDisabled
        }

        syncStatus = .syncing

        do {
            try await syncManager.performSync()
            syncStatus = .completed
        } catch {
            syncStatus = .failed(error)
            throw error
        }
    }

    // MARK: - Private Helpers

    private func setupAutoSync() {
        syncTimer = Timer.scheduledTimer(
            withTimeInterval: configuration.syncInterval,
            repeats: true
        ) { [weak self] _ in
            Task { @MainActor [weak self] in
                try? await self?.syncNow()
            }
        }
    }

    private func stopAutoSync() {
        syncTimer?.invalidate()
        syncTimer = nil
    }

    private func saveAuthToken(_ token: AuthToken) {
        // Save to Keychain in production
        if let data = try? JSONEncoder().encode(token) {
            UserDefaults.standard.set(data, forKey: "auth_token")
        }
    }

    private func loadAuthToken() {
        if let data = UserDefaults.standard.data(forKey: "auth_token"),
           let token = try? JSONDecoder().decode(AuthToken.self, from: data) {
            self.authToken = token
            self.isAuthenticated = true
        }
    }

    private func clearAuthToken() {
        UserDefaults.standard.removeObject(forKey: "auth_token")
    }

    // MARK: - Internal API

    func makeRequest<T: Decodable>(
        _ endpoint: String,
        method: String = "GET",
        body: Data? = nil
    ) async throws -> T {
        guard let url = URL(string: endpoint, relativeTo: configuration.baseURL) else {
            throw DatenError.invalidResponse
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")

        if let token = authToken {
            request.addValue("Bearer \(token.accessToken)", forHTTPHeaderField: "Authorization")
        } else if let apiKey = configuration.apiKey {
            request.addValue(apiKey, forHTTPHeaderField: "X-API-Key")
        }

        if let body = body {
            request.httpBody = body
        }

        let (data, response) = try await urlSession.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw DatenError.invalidResponse
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try decoder.decode(T.self, from: data)
    }
}

// MARK: - Authentication Service

class AuthenticationService {
    private weak var client: DatenClient?

    init(client: DatenClient) {
        self.client = client
    }

    func signIn(credentials: AuthCredentials) async throws -> AuthToken {
        let body = [
            "username": credentials.username,
            "password": credentials.password
        ]

        let data = try JSONEncoder().encode(body)
        return try await client!.makeRequest("/auth/login", method: "POST", body: data)
    }

    func refreshToken(refreshToken: String) async throws -> AuthToken {
        let body = ["refresh_token": refreshToken]
        let data = try JSONEncoder().encode(body)
        return try await client!.makeRequest("/auth/refresh", method: "POST", body: data)
    }
}

// MARK: - Document Service

class DocumentService {
    private weak var client: DatenClient?

    init(client: DatenClient) {
        self.client = client
    }

    func fetchDocuments() async throws -> [Document] {
        return try await client!.makeRequest("/documents")
    }

    func createDocument(_ document: Document) async throws -> Document {
        let data = try JSONEncoder().encode(document)
        return try await client!.makeRequest("/documents", method: "POST", body: data)
    }

    func updateDocument(_ document: Document) async throws -> Document {
        let data = try JSONEncoder().encode(document)
        return try await client!.makeRequest("/documents/\(document.id)", method: "PUT", body: data)
    }

    func deleteDocument(id: String) async throws {
        let _: [String: String] = try await client!.makeRequest("/documents/\(id)", method: "DELETE")
    }
}

// MARK: - Sync Manager

enum SyncOperation {
    case create(Document)
    case update(Document)
    case delete(String)
}

@MainActor
class SyncManager {
    private weak var client: DatenClient?
    private var pendingOperations: [SyncOperation] = []

    init(client: DatenClient) {
        self.client = client
    }

    func queueOperation(_ operation: SyncOperation) {
        pendingOperations.append(operation)
    }

    func performSync() async throws {
        guard let client = client else { return }

        for operation in pendingOperations {
            switch operation {
            case .create(let document):
                _ = try await client.documentService.createDocument(document)
            case .update(let document):
                _ = try await client.documentService.updateDocument(document)
            case .delete(let id):
                try await client.documentService.deleteDocument(id: id)
            }
        }

        pendingOperations.removeAll()
    }
}

// MARK: - Biometric Service

class BiometricAuthService {
    private let context = LAContext()

    func canAuthenticateWithBiometrics() -> Bool {
        var error: NSError?
        return context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error)
    }

    func authenticate(reason: String) async throws -> Bool {
        guard canAuthenticateWithBiometrics() else {
            throw DatenError.biometricAuthFailed
        }

        return try await context.evaluatePolicy(
            .deviceOwnerAuthenticationWithBiometrics,
            localizedReason: reason
        )
    }
}

// MARK: - SwiftUI Preview Support

#if DEBUG
extension DatenClient {
    static var preview: DatenClient {
        let config = DatenConfiguration(
            baseURL: URL(string: "https://api.example.com")!,
            enableOfflineMode: false
        )
        return DatenClient(configuration: config)
    }
}
#endif
